File: core/src/main/java/org/evergreen_ils/android/Log.java
/*
 * Copyright (c) 2020 Kenneth H. Cox
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

package org.evergreen_ils.android;

/** private logging class that allows substituting different behaviors,
 * Created by kenstir on 12/9/2015.
 */
public class Log {
    public static LogProvider provider = new AndroidLogProvider();

    // defining these statics here allows me to unit test code that logs / calls analytics
    // values from https://developer.android.com/reference/android/util/Log
    public static final int DEBUG = 3;
    public static final int WARN = 5;

    public static void setProvider(LogProvider _provider) {
        provider = _provider;
    }
    public static String getLogBuffer() {
        if (provider != null) return provider.getLogBuffer();
        return null;
    }

    public static void d(String TAG, String msg) {
        if (provider != null) provider.d(TAG, msg);
    }
    public static void d(String TAG, String msg, Throwable tr) {
        if (provider != null) provider.d(TAG, msg, tr);
    }
    public static void i(String TAG, String msg) {
        if (provider != null) provider.i(TAG, msg);
    }
    public static void w(String TAG, String msg) {
        if (provider != null) provider.w(TAG, msg);
    }
    public static void w(String TAG, String msg, Throwable tr) {
        if (provider != null) provider.w(TAG, msg, tr);
    }

    public static long logElapsedTime(String TAG, long start_ms, String s) {
        long now_ms = System.currentTimeMillis();
        if (provider != null) provider.d(TAG, String.format("%3dms elapsed: %s", now_ms - start_ms, s));
        return now_ms;
    }
}

File: core/src/main/java/net/kenstir/logging/Log.kt
/*
 * Copyright (c) 2025 Kenneth H. Cox
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <https://www.gnu.org/licenses/>.
 */
package net.kenstir.logging

import android.annotation.SuppressLint

/**
 * Logger that automatically uses Android logging when available, stdout when not.
 *
 * This allows liberal use of logging while still allowing unit tests to run without an emulator.
 */
object Log {
    // defining these statics here allows me to unit test code that logs / calls analytics
    // values from https://developer.android.com/reference/android/util/Log
    const val VERBOSE: Int = 2
    const val DEBUG: Int = 3
    const val WARN: Int = 5

    const val TAG_ASYNC = "async"
    const val TAG_EXTENSIONS = "extensions"
    const val TAG_FCM = "fcm"
    const val TAG_PERM = "perm"

    val provider: LogProvider by lazy { getLogProvider() }
    var level: Int = DEBUG

    fun v(tag: String, msg: String) {
        if (level <= VERBOSE) provider.v(tag, msg)
    }

    @JvmStatic
    fun d(tag: String, msg: String) {
        if (level <= DEBUG) provider.d(tag, msg)
    }

    @JvmStatic
    fun d(tag: String, msg: String, tr: Throwable?) {
        if (level <= DEBUG) provider.d(tag, msg, tr)
    }

    @JvmStatic
    @SuppressLint("DefaultLocale")
    fun logElapsedTime(tag: String, startTime: Long, s: String): Long {
        val now = System.currentTimeMillis()
        provider.d(tag, String.format("%3dms: %s", now - startTime, s))
        return now
    }

    private fun getLogProvider(): LogProvider {
        return AndroidLogProvider.create() ?: StdoutLogProvider()
    }
}
